lib/async-emitter.js

```diff
@@ -0,0 +1,110 @@
+'use strict';
+
+const { iter } = require('@metarhia/common');
+
+class AsyncEmitter {
+  constructor() {
+    this.events = new Map();
+    this.wrappers = new Map();
+  }
+
+  // Add listener
+  //   name <string> event name
+  //   fn <Function> listener
+  on(name, fn) {
+    let event = this.events.get(name);
+    if (!event) {
+      event = new Set();
+      this.events.set(name, event);
+    }
+    event.add(fn);
+  }
+
+  // Add listener
+  //   name <string> event name
+  //   fn <Function> listener
+  // Returns: <Promise> | <undefined>
+  once(name, fn) {
+    if (fn === undefined) {
+      return new Promise(resolve => {
+        this.once(name, resolve);
+      });
+    }
+    const wrapper = (...args) => {
+      this.remove(name, fn);
```

Comment:
This has a side effect of removing the other listeners for the event added via `on()` method, so for example:
```js
e.once('eventName', listener);

// ... somewhere else ...
e.on('eventName', listener);
```
Here both listeners will be removed after the first time the event is emitted, IIUC.

Reply:
Yes, have found this problem too, trying to solve